### ISSUE TITLE
False positives of ModuleInitialize

### ISSUE THREAD

**Issue by PeterCamilleri** [troessner/reek#1137]
Two closely related issues with the following fragment of code:

```
module Mysh
  def self.reset_host
    exec_class = Class.new do
      #Set up a new execution environment
      #<br>Endemic Code Smells -- False positive.
      #* :reek:ModuleInitialize
      def initialize
        #Code deleted for brevity.
      end

      #Balance of class definition omitted.
    end

    #Balance of reset_host method deleted for brevity.
end
```
(1) This generates the following warning:

```
lib/mysh/expression.rb -- 1 warning:
  [9]:ModuleInitialize: Mysh has initialize method [https://github.com/troessner/reek/blob/master/docs/Module-Initialize.md]
1 total warning
```
even though it is a class initialize method and not a module initialiaze method.

(2) The :reek:ModuleInitialize directive is ignored. The only way to silence this report is to turn it off in the project's reek config file. I think this is a poor option since it prevents the detection of legitimate code smells.

**Comment by troessner** [comment]
Yeah, that's not something we can detect at the moment.
The AST for dynamically creating classes is vastly different than for creating it "regularly" (if there is such a thing as "regular" with Ruby).

```
 ruby-parse -e 'module M; def m; Class.new; end; end'
(module
  (const nil :M)
  (def :m
    (args)
    (send
      (const nil :Class) :new)))
```

vs.

```
ruby-parse -e 'module M; class C; end; end'
(module
  (const nil :M)
  (class
    (const nil :C) nil nil))
```

Or in other words, we'd have to detect `Class.new` as well. That might open up a few more edge cases as well. WDYT @mvz ?

>> (2) The :reek:ModuleInitialize directive is ignored. The only way to silence this report is to turn it off in the project's reek config file. I think this is a poor option since it prevents the detection of legitimate code smells.

You have to put the comment above the module like this:

```Ruby
      # :reek:ModuleInitialize
      module Alfa
        def initialize; end
      end
```

That's because the smell detector works on module level, not on method level. 
Something we failed to document properly in our [detector docs](https://github.com/troessner/reek/blob/master/docs/Module-Initialize.md), I'll update this and all similar pages over the next days.

**Comment by mvz** [comment]
We are actually detecting `Class.new` (and `Struct.new`) in a few cases. It would be nice if we could consistently handle them the same way as `class Foo`. It will be tricky, since we assume the 'regular' way in quite a few places.

OTOH, we should probably not even look inside `def self.reset_host` at all in this case.

**Comment by chastell** [comment]
This should be fixed by #1287.
